core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,643 @@
+						  hashdata->ctx.length));
+		caam_desc_add_ptr(desc, hashdata->ctx.paddr);
+
+		cache_operation(TEE_CACHEINVALIDATE, hashdata->ctx.data,
+				hashdata->ctx.length);
+		HASH_DUMPBUF("CTX", hashdata->ctx.data, hashdata->ctx.length);
+		hashdata->ctx.length = 0;
+	} else {
+		HASH_TRACE("Init/Final Operation");
+		caam_desc_add_word(desc, HASH_INITFINAL(alg->type));
+	}
+
+	HASH_DUMPBUF("Temporary Block", hashdata->blockbuf.buf.data,
+		     hashdata->blockbuf.filled);
+	caam_desc_add_word(desc, FIFO_LD_EXT(CLASS_2, MSG, LAST_C2));
+	caam_desc_add_ptr(desc, hashdata->blockbuf.buf.paddr);
+	caam_desc_add_word(desc, hashdata->blockbuf.filled);
+	cache_operation(TEE_CACHECLEAN, hashdata->blockbuf.buf.data,
+			hashdata->blockbuf.filled);
+	hashdata->blockbuf.filled = 0;
+
+	/* Save the final digest */
+	caam_desc_add_word(desc, ST_NOIMM(CLASS_2, REG_CTX, alg->size_digest));
+	caam_desc_add_ptr(desc, digest_align.paddr);
+
+	HASH_DUMPDESC(desc);
+
+	jobctx.desc = desc;
+
+	if (digest_align.nocache == 0)
+		cache_operation(TEE_CACHEFLUSH, digest_align.data,
+				alg->size_digest);
+
+	retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+	if (retstatus == CAAM_NO_ERROR) {
+		/* Ensure that hash data are correct in cache */
+		if (digest_align.nocache == 0)
+			cache_operation(TEE_CACHEINVALIDATE, digest_align.data,
+					alg->size_digest);
+
+		ret = TEE_SUCCESS;
+		if (realloc)
+			memcpy(digest, digest_align.data, len);
+
+		HASH_DUMPBUF("Digest", digest_align.data,
+			     (size_t)alg->size_digest);
+	} else {
+		HASH_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
+		ret = job_status_to_tee_result(jobctx.status);
+	}
+
+exit_final:
+	if (realloc == 1)
+		caam_free_buf(&digest_align);
+
+	return ret;
+}
+
+/*
+ * Copy Sofware Hashing Context
+ *
+ * @src_ctx  Reference the context source
+ * @dst_ctx  [out] Reference the context destination
+ */
+static void do_copy_state(struct crypto_hash_ctx *dst_ctx,
+			  struct crypto_hash_ctx *src_ctx)
+{
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
+	struct crypto_hash *hash_src = to_hash_ctx(src_ctx);
+	struct crypto_hash *hash_dst = to_hash_ctx(dst_ctx);
+	struct hashdata *dst = hash_dst->ctx;
+	struct hashdata *src = hash_src->ctx;
+
+	HASH_TRACE("Copy State context (0x%" PRIxPTR ") to (0x%" PRIxPTR ")",
+		   (uintptr_t)src, (uintptr_t)dst);
+
+	if (!dst || !src)
+		return;
+
+	dst->alg = src->alg;
+
+	if (!dst->ctx.data) {
+		retstatus = do_allocate_intern(dst);
+		if (retstatus != CAAM_NO_ERROR)
+			return;
+	}
+
+	memcpy(dst->ctx.data, src->ctx.data, src->ctx.length);
+	dst->ctx.length = src->ctx.length;
+	cache_operation(TEE_CACHECLEAN, dst->ctx.data, dst->ctx.length);
+
+	if (src->blockbuf.filled) {
+		struct caambuf srcdata = { .data = src->blockbuf.buf.data,
+					   .length = src->blockbuf.filled };
+
+		caam_cpy_block_src(&dst->blockbuf, &srcdata, 0);
+	}
+}
+
+/*
+ * Registration of the HASH Driver
+ */
+static const struct crypto_hash_ops hash_ops = {
+	.init = &do_init,
```

Comment:
`&` in front of the function name is redundant.

Reply:
ok